### ISSUE TITLE
Perform stack probe on task workers/local sync to ensure stack is valid for dispatches on Windows.

### ISSUE THREAD

**Issue by benvanik** [iree-org/iree#16294]
We currently rely on luck to ensure that the stack guard pages have been appropriately unlocked for use by dispatches that use more than a page of stack space. We need each caller of embedded ELF dispatch functions to call `_chkstk` to probe out to some boundary (64K min, but could go to 128K for margin) to avoid cases where a dispatch overruns the currently unlocked pages. We want to amortize the probing so that each dispatch workgroup does not perform the work. We control execution of the task workers and can perform the probing once on worker creation. For the sync driver we don't own the calling thread and may move across threads between execution of command buffers (but never within) and can perform the probe in command buffer begin to then ensure all dispatches within the command buffer are good.

Calling `_chkstk` can be tricky as it does not follow a standard calling convention because it's meant to be emitted as part of a function prologue. We'll need a .asm on x86-64 but can use inline asm on x86-32. We could make this an implementation detail of the embedded ELF loader by having a loader vtable for performing thread prep that's called by the drivers.

**Comment by LLITCHEV** [comment]
First, i thought that we would be crashing almost all the time on Windows, looking at the codegen and how much stack space we use - we often go over the 8k stack size (which is the default stack size on 64 bit x86 Windows OS, generated by the compilers and the automatically committed memory). Then, I remembered vaguely (from about 10-ish years ago when I was writing experimental LLVM Jit for the CLR for no-MSFT OSs) that LLVM handled this in the prolog generation for Windows. Like Phoenix Compiler, LLVM has a precise stack size calculation, so it knows the exact size of the stack for each frame. I thought for the second that it might have been removed, but it looks it is still there. Since we use LLVM codegen for Windows, this should be already handled, if I'm understanding the issue described here properly - https://github.com/llvm-mirror/llvm/blob/cd5b3fa3e75b54b04a3ef6ce5229f1c6b0d6b2b8/lib/Target/X86/X86FrameLowering.cpp#L1147C3-L1172C1
Or am I misunderstanding completely everything?

**Comment by benvanik** [comment]
Close! :)
We generate portable ELF files that we use on all platforms - these ELFs cannot make syscalls as they aren't portable and cannot assume anything about the hosting environment as there may not even be running in one (bare metal). Thus, the caller of our generated code needs to perform anything that would normally be done as part of platform-dependent logic within the generated code, such as _chkstk. We only skate by today because enough of the stack for most of our programs has been committed but if we start to lean more on larger stacks we need to deterministically control the amount we've committed by issuing the _chkstk ourselves.

see: https://github.com/openxla/iree/tree/main/runtime/src/iree/hal/local/elf

**Comment by MaheshRavishankar** [comment]
We statically check that the stack allocation is bounded? Maybe the check is off somewhere if you are seeing issues.

**Comment by benvanik** [comment]
Can't we only do that with statically shaped allocations? Or do we also disallow dynamic allocations?

(even 32KB may trip the guard pages - we're just lucky today in the common cases where they're already committed - the rule is >4KB on x86 or >8KB on x64 need a _chkstk)

**Comment by benvanik** [comment]
from your comment in #16249 sounds like it prevents unbounded allocations - neato. Then this is mostly a CYA for conforming to the platform requirements vs something we need ASAP.

**Comment by MaheshRavishankar** [comment]
> from your comment in #16249 sounds like it prevents unbounded allocations - neato. Then this is mostly a CYA for conforming to the platform requirements vs something we need ASAP.

Yes, no more unbounded allocations on CPU side. That is checked statically.
